xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,550 @@
+                 result_types: Sequence[Attribute],
+                 successors: Sequence[IBlock],
+                 regions: Sequence[IRegion]) -> None:
+        object.__setattr__(self, "_op_data", op_data)
+        object.__setattr__(self, "operands", IList(operands))
+        for operand in operands:
+            operand._add_user(self) # type: ignore
+        object.__setattr__(
+            self, "results",
+            IList([
+                IResult(type, IList([]), self, idx)
+                for idx, type in enumerate(result_types)
+            ]))
+        object.__setattr__(self, "successors", IList(successors))
+        object.__setattr__(self, "regions", IList(regions))
+
+        self.operands.freeze()
+        self.results.freeze()
+        self.successors.freeze()
+        self.regions.freeze()
+
+    @classmethod
+    def get(cls, name: str, op_type: type[Operation],
+            operands: Sequence[ISSAValue], result_types: Sequence[Attribute],
+            attributes: dict[str, Attribute], successors: Sequence[IBlock],
+            regions: Sequence[IRegion]) -> IOp:
+        return cls(OpData(name, op_type, attributes), operands, result_types,
+                   successors, regions)
+
+    def __hash__(self) -> int:
+        return hash(id(self))
+
+    def __eq__(self, __o: object) -> bool:
+        return self is __o
+
+    @property
+    def name(self) -> str:
+        return self._op_data.name
+
+    @property
+    def op_type(self) -> Type[Operation]:
+        return self._op_data.op_type
+
+    @property
+    def attributes(self) -> dict[str, Attribute]:
+        return self._op_data.attributes
+
+    @property
+    def result(self) -> IResult | None:
+        if len(self.results) > 0:
+            return self.results[0]
+        return None
+
+    @property
+    def region(self) -> IRegion | None:
+        if len(self.regions) > 0:
+            return self.regions[0]
+        return None
+
+    @property
+    def result_types(self) -> List[Attribute]:
+        return [result.typ for result in self.results]
+
+    def get_mutable_copy(
+            self,
+            value_mapping: Optional[dict[ISSAValue, SSAValue]] = None,
+            block_mapping: Optional[dict[IBlock, Block]] = None) -> Operation:
+        """
+        Returns a mutable operation that is a copy of this immutable operation.
+        The value_mapping and block_mapping are used to map already known correspondings
+        of immutable values to mutable values and immutable blocks to mutable blocks.
+        """
+        if value_mapping is None:
+            value_mapping = {}
+        if block_mapping is None:
+            block_mapping = {}
+
+        mutable_operands: List[SSAValue] = []
+        for operand in self.operands:
+            if operand in value_mapping:
+                mutable_operands.append(value_mapping[operand])
+            else:
+                print(f"ERROR: op {self.name} uses SSAValue before definition")
```

Comment:
We should fix this at some point, but I believe this doesn't work when blocks are not in order (which technically could happen)

Reply:
Also, we should probably return an exception rather than just printing something.

Reply:
> We should fix this at some point, but I believe this doesn't work when blocks are not in order (which technically could happen)

Ah, you mean when an op uses the result from an op in a block below, but which dominates the current block?

Reply:
> Also, we should probably return an exception rather than just printing something.

For debugging it is incredibly useful to still form a complete (although with ill formed references) program and print it so the error can be seen in context. But I agree, we should come up with a more principled way than this.

The problem is, the immutable IR can not be printed on its own, without converting to mutable first.

Reply:
Maybe there exists a way to combine this with our inlined exception prints? (However, that will first have to be readded to the printer, I think).

Reply:
Is there progress or a pathway to achieve the behavior you are after or should I leave it like this?

I would like to report an error here, but still continue with the process such that the error can be examined in context. 
I could just move this to throw an exception, but this is so so valuable to me. I would then prefer to have a `DEVMODE` flag or something to enable continuing even though there were errors.

Reply:
I am fine with leaving as is. @math-fehr ?